### PR TITLE
Add a design doc for admission webhook bootstrapping

### PR DESCRIPTION
**Pull request by caesarxuchao**
Originally distributed as a [google doc](https://docs.google.com/document/d/1pw6FyobY3pVxfWYwmAFuF5WJpvjfq-M3Ciz996kblVc/edit?ts=59f1d824#).

The idea is from an Oct. 23 meeting with lavalamp, cheftako, deads2k, smarterclayton, liggitt. It also draws inspiration from #1191.

cc @kubernetes/sig-api-machinery-api-reviews  @lavalamp @deads2k

### CODE REVIEW

File: contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Webhook Bootstrapping
```
Comment (lavalamp) [inline contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md]: On file location: did I make a mistake putting the larger design doc in sig-api-machinery/design-docs?

File: contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+## Background
+Admission webhook is a feature that dynamically extends Kubernetes admission
```
Comment (lavalamp) [inline contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md]: You can link back to the larger design doc if you wish.

File: contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+We will add a new field `NamespaceSelectorq to it:
+
+```golang
+type ExternalAdmissionHook struct {
+    Name string
+    ClientConfig AdmissionHookClientConfig
+    Rules []RuleWithOperations
+    FailurePolicy *FailurePolicyType
+    // Only objects in matching namespaces are subjected to this webhook.
+    // LabelSelector.MatchExpressions allows reverse selection, so you can use this
```
Comment (lavalamp) [inline contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md]: maybe "... allows exclusive as well as inclusive matching, so you ..."

**Comment by lavalamp** [comment]
tiny nits, LGTM from me -- will give David a chance to see it

File: contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+We add a selector to the admission webhook configuration, which will be compared
+to the labels of namespaces. Only objects in the matching namespaces are
+subjected to the webhook admission. A cluster admin will want to exempt these
+namespaces from webhooks:
+- Namespaces where this webhook and other webhooks are deployed in;
+- Namespaces where core system components are deployed in.
+
+## API Changes
+`ExternalAdmissionHook` is the dynamic configuration API of an admission webhook.
+We will add a new field `NamespaceSelectorq to it:
```
Comment (smarterclayton) [inline contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md]: typo

File: contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    // See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ for more examples of label selectors.
+    NamespaceSelector *metav1.LabelSelector
+}
+```
+
+## Guidelines on namespace labeling
+The mechanism depends on cluster admin properly labelling the namespaces. We
+will provide guidelines on the labelling scheme. One suggestion is labelling
+namespaces with runlevels. The design of runlevels is out of the scope of this
+document (tracking issue), a strawman runlevel scheme is:
```
Comment (smarterclayton) [inline contributors/design-proposals/api-machinery/admission-webhook-bootstrapping.md]: Don't forget to add the tracking issue

**Comment by caesarxuchao** [comment]
Addressed the comments.

**Comment by deads2k** [comment]
lgtm

**Pull request merged**